### PR TITLE
Add WASM renderer

### PR DESCRIPTION
**Pull request by bcspragu**
I'm still working on the API and the formatting and everything, but I have an example algo [1] that compiles to WASM, so now I need to wire them together.

Still lots missing at this point

[1] https://git.sr.ht/~bsprague/recurse/tree/main/item/rapidriter/simpletest

### CODE REVIEW

File: web/core/templates/index.html [hunk 1]
```diff
@@ -66,6 +66,28 @@ <h3>ALL SHOWS</h3>
...
+
+  add a WASM show below! just write some WASM!!
+
+  <i>HOW TO WASM:</i>
+  <pre>
+    - Paste your <a href="https://developer.mozilla.org/en-US/docs/WebAssembly/Understanding_the_text_format">WebAssembly Text (WAT)</a> file here
+    - have a `next_frame` function that returns a 96x38 array of booleans, or an empty array to indicate you're done
+  </pre>
+
+  feel free to <a href="https://recurse.zulipchat.com/#narrow/dm/7864">ping @jessechen</a> or <a href="https://github.com/jessechen/bubbletea">make a pull request</a> to add more!!
```
Comment (gregsadetsky) [inline web/core/templates/index.html:82]: extremely small comment, this block (re: "ping Jesse") is related to the "add text show" above, if you want to move that above

**Comment by gregsadetsky** [comment]
really really really really amazing!!!!!!!!! extremely excited to see it live :-)

anything I can help with? feel free to merge if you want to try it out :-) does it work for you locally?

File: web/core/templates/index.html (hunk 1 above)
Comment (bcspragu) [reply web/core/templates/index.html:82]: Indeed! Mostly a copy paste error, will definitely fix

**Comment by bcspragu** [comment]
> anything I can help with? feel free to merge if you want to try it out :-) does it work for you locally?

Definitely not merge-ready :) But maybe tomorrow! I need to figure out how to read the data out of WASM first, and then figure out how to wire it all up

**Comment by nathanleiby** [comment]
> I'm still working on the API and the formatting and everything, but I have an example algo [1] that compiles to WASM, so now I need to wire them together.
> 
> Still lots missing at this point
> 
> [1] https://git.sr.ht/~bsprague/recurse/tree/main/item/rediff

Very cool! 

I've also been learning more about WASM today (as a higher level consumer of wasm-bindgen in rust, so far). I would be interested in building a viz here.

**Comment by bcspragu** [comment]
> I've also been learning more about WASM today (as a higher level consumer of wasm-bindgen in rust, so far). I would be interested in building a viz here.

Haha I saw your check-in from yesterday and was going to reach out! Would love to pair on that, the part I'm struggling with is getting the structured data out from WASM on the other end, and solving that is required to visualize it

**Comment by bcspragu** [comment]
Okay, just pushed the latest and greatest, and I think it's ready for review. Latest update adds:

- Documentation for running everything in Docker
- A Dockerfile for the WASM rendererer to make it Disco-able
- Frame limiting (theoretically)
  - In practice, using the terminal logging as a guide, it looks super janky and there might be something wrong with my methodology

**Comment by gregsadetsky** [comment]
So so so grateful!!!

I'll merge and deploy and let you know

And then try converting your simpletest into a wat and running that..? If you have a web assembly text program handy feel free to send it

Thanks again!!

**Comment by bcspragu** [comment]
> And then try converting your simpletest into a wat and running that..? If you have a web assembly text program handy feel free to send it

Was exactly my thought process, and indeed I do!

<details>
<summary>simpletest.wat</summary>

```
(module
  (type (;0;) (func (param i32) (result i32)))
  (func (;0;) (type 0) (param i32) (result i32)
    local.get 0
    i32.const 100
    i32.gt_u)
  (func (;1;) (type 0) (param i32) (result i32)
    (local i32)
    global.get 0
    i32.const 464
    i32.sub
    local.tee 1
    global.set 0
    local.get 1
    i32.const 464
    i32.add
    global.set 0
    i32.const 0
    local.get 1
    i32.const 8
    i32.add
    local.get 0
    i32.const 100
    i32.gt_u
    select)
  (table (;0;) 1 1 funcref)
  (memory (;0;) 16)
  (global (;0;) (mut i32) (i32.const 1048576))
  (global (;1;) i32 (i32.const 1048576))
  (global (;2;) i32 (i32.const 1048576))
  (export "memory" (memory 0))
  (export "is_done" (func 0))
  (export "next_frame" (func 1))
  (export "__data_end" (global 1))
  (export "__heap_base" (global 2)))
```

</details>

This should, in theory, light up the first 20 LEDS, then the second 20 LEDs next frame, etc, etc, for 100 frames. I also realized I was never sending the `end` event over SSE, so that exists now. The output is still unexpectedly unintelligible in the terminal output, but we'll see what it looks like on the RR. Hopefully no magic smoke, but honestly that would be kind of an impressive outcome too

**Pull request merged**